### PR TITLE
utils/small_vector: refactor expansion condition in reserve*()

### PR DESCRIPTION
**Pull request by lersek**
~~~
utils/small_vector: refactor expansion condition in reserve*()

Rewrite

  _begin + n > _capacity_end

as

  n > _capacity_end - _begin

and then as

  n > capacity()

for two reasons:

- The last form is easier to read than the first form.

- Per N4950 (the final C++23 working draft), [expr.add] paragraph 4, the
  expression

    _begin + n                            (i.e., P + J)

  is defined only if

    0 ≤ 0 + n ≤ _capacity_end - _begin    (i.e., 0 ≤ i + j ≤ n)

  equivalently, only if

    _begin ≤ _begin + n ≤ _capacity_end

  Therefore, the expression

    _begin + n

  invokes undefined behavior exactly when we'd expect our check

    _begin + n > _capacity_end

  to evaluate to true.

gcc and clang have been aggressively equating undefined behavior to "never
happens"; let's prevent that here.
~~~

No backports needed.

### CODE REVIEW

**Comment by tchaikov** [comment]
@lersek hi Laszlo, thank you for fixing the build of tests. i think your change looks quite like #21183. could you help review them?

File: utils/small_vector.hh [hunk 1]
```diff
@@ -105,7 +105,7 @@ private:
     }
 
     void reserve_at_least(size_t n) {
-        if (__builtin_expect(_begin + n > _capacity_end, false)) {
+        if (__builtin_expect(n > capacity(), false)) {
```
Comment (tchaikov) [inline utils/small_vector.hh:108]: > removes the (theoretical) undefined behavior in "_begin+n"

i think this does not hold, as `_begin` is a plain pointer, so unless your concern is that the integer overflow, the behavior of `_begin + n` should be defined.

**Comment by lersek** [comment]
@tchaikov 

> @lersek hi Laszlo, thank you for fixing the build of tests. i think your change looks quite like #21183. could you help review it?

Yes, I guess I should have checked first; your #21183 preempts the second patch here (the patch bodies are identical). Thanks!

File: utils/small_vector.hh (hunk 1 above)
Comment (lersek) [reply utils/small_vector.hh:108]: No, my concern is not integer overflow. The pre-patch code is undefined behavior (strictly speaking) for a different reason.

Using the `+` and `-` operators, you can **form** (not dereference!) only such pointers that point to an element in an array, or one past the array (i.e., point just behind the array). The *one past* terminology is from the C standard; the C++ standard uses the term *possibly-hypothetical*.

(For this discussion, a standalone object `x` is considered a single-element array `x[1]`; therefore the valid pointers to evaluate are `&x` and `&x + 1`. From these, only `&x` can be dereferenced; `*&x` is fine, `*(&x + 1)` is UB. But my point here is not dereferencing -- my point is evaluating. Namely, that `&x + 2` cannot even be *evaluated* (not dereferenced), strictly speaking.)

See `[expr.add]` in `n4950.pdf`, paragraph 4 (sub-paragraphs 4.2 and 4.3).

Here, `_begin` points to an array with `_capacity_end - _begin` elements. Therefore, we are allowed to evaluate:
- `_begin + 0`,
- `_begin + 1`,
- ...
- `_begin + (_capacity_end - _begin)`, which can also be written simply as `_capacity_end`.

(All but the last one can also be dereferenced, but that's beside the point here.)

The point is that, using the expression `_begin + n`, you cannot *form* a pointer that points past `_capacity_end`.

In other words, exactly in the case when we expected the comparison `_begin + n > _capacity_end` to evaluate to `1`, the addition `_begin + n` itself is already undefined behavior. Per the language standard, you cannot rely on an array bounds check in this form. You have to write it as `n > _capacity_end - _begin`.

(And to that pointer subtraction form, the next paragraph of the standard applies; paragraph 5.)

Comment (dawmd) [reply utils/small_vector.hh:108]: Could you elaborate on the problem in the commit message, even if briefly? It will make what kind of UB is being solved clearer.

Comment (lersek) [reply utils/small_vector.hh:108]: OK, I'll do that in v2. Thanks!

**Comment by lersek** [comment]
v2:
- drop second patch [Kefu]
- rewrite commit message on first patch [Dawid]

File: utils/small_vector.hh (hunk 1 above)
Comment (lersek) [reply utils/small_vector.hh:108]: done in v2

**Review by dawmd — APPROVED** [review]
LGTM

File: utils/small_vector.hh (hunk 1 above)
Comment (tchaikov) [reply utils/small_vector.hh:108]: i see. so there are two cases:

- `_begin` is a plain pointer, it is assigned with the return value of `::aligned_alloc()` 
- `_begin` points to an element in `storage`. which is an array

but we don't intend to differentiate the second case from the first case.

**Review by tchaikov — APPROVED** [review]
lgtm.

**Comment by tchaikov** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12634/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12634/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1287/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12375/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4352/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12634/)
> #### Failed Tests (1/36274):
> 
>     * [test_apply_to_incomplete_respects_continuity](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12634/testReport/junit/boost/mvcc_test/test_apply_to_incomplete_respects_continuity) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_apply_to_incomplete_respects_continuity)
> 
> 
#13757

File: utils/small_vector.hh (hunk 1 above)
Comment (lersek) [reply utils/small_vector.hh:108]: Right; from the pointer arithmetic angle, there is no difference between them.

The original idea behind this can be found in the C89 Rationale:

> Consider a hypothetical segmented architecture, on which pointers comprise a segment descriptor and an offset. Suppose that segments are relatively small, so that large arrays are allocated in multiple segments. While the segments are valid (allocated, mapped to real memory), the hardware, operating system, or C implementation can make these multiple segments behave like a single object: pointer arithmetic and relational operators use the defined mapping to impose the proper order on the elements of the array. Once the memory is deallocated, the mapping is no longer guaranteed to exist; use of the segment descriptor might now cause an exception, or the hardware addressing logic might return meaningless data.

So in theory even just evaluating a pointer may involve some "internal dereferencing" (see the "segment descriptor" term). Therefore we can only evaluate pointers that point to "valid address space" (and null pointers, and pointers to just past the end of an array). In this sense, there is no difference whether the "destination address space" of `_begin` is valid because `_begin` points to `storage`, or to an area allocated with `::aligned_alloc()`. What matters is that the "descriptor tables" that the `_begin + n` pointer arithmetic internally dereferences *exist*.

Obviously the environments we care about are not such "segmented architectures"; segmentation in x86_64 long mode is not used for address translation, and paging is hidden by the OS and the processor on x86_64 and arm64; the pointers that userspace sees are "flat" (assuming no tagged pointers, but I don't think we use those). Therefore in practice you will never see `_begin + n` cause problems -- as long as the compiler translates it to a simple multiply+add (or equivalent) instruction, that is. The problem is that compilers have been getting more aggressive at exploiting behavior that's officially undefined, for optimization purposes. So it's not a "dangling segment descriptor" that we need to protect against; the adversary is now the compiler (or language runtime). A "sufficiently smart" compiler could optimize `_begin + n > _capacity_end` to constant `0` or `1` *today*, and per the language standard, it would be in its right to do so.

I apologize if this does not seem like a practical consideration; that's why my other argument went like, "`n > capacity()` is easier to read" :)

**Comment by lersek** [comment]
> retriggering the build.

Thank you Kefu!

File: utils/small_vector.hh (hunk 1 above)
Comment (tchaikov) [reply utils/small_vector.hh:108]: thank you Laszlo for the detailed explanation! this indeed helped me understand why the standard has this weird consideration. i was pulling my hairs out yesterday while reading the standard draft, but switched away after about 10 minutes without digging into it.
